blast/ncbi/doc/man/asndhuff.1
.TH ASNDHUFF 1 2002-08-30 NCBI "NCBI Tools User's Manual"
.SH NAME
asndhuff \- decompress compressed ASN.1 (CASN) files
.SH SYNOPSIS
.B asndhuff
[\|\fB\-\fP\|]
\fB\-i\fP\ \fIfilename\fP
[\|\fB\-o\fP\ \fIfilename\fP\|]
.SH DESCRIPTION
\fBasndhuff\fP decompresses compressed ASN.1 (CASN) files.
.SH OPTIONS
A summary of options is included below.
.TP
\fB\-\fP
Print usage message
.TP
\fB\-i\fP\ \fIfilename\fP
Read input from \fIfilename\fP
.TP
\fB\-o\fP\ \fIfilename\fP
Write output to \fIfilename\fP (stdout by default)
.SH AUTHOR
The National Center for Biotechnology Information.
.SH SEE ALSO
.BR asn2asn (1),
.BR asn2ff (1),
.BR asn2gb (1),
.BR asn2xml (1)


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

blast/demo/asndhuff.c
/* asndhuff.c
* ===========================================================================
*
*                            PUBLIC DOMAIN NOTICE                          
*               National Center for Biotechnology Information
*                                                                          
*  This software/database is a "United States Government Work" under the   
*  terms of the United States Copyright Act.  It was written as part of    
*  the author's official duties as a United States Government employee and 
*  thus cannot be copyrighted.  This software/database is freely available 
*  to the public for use. The National Library of Medicine and the U.S.    
*  Government have not placed any restriction on its use or reproduction.  
*                                                                          
*  Although all reasonable efforts have been taken to ensure the accuracy  
*  and reliability of the software and data, the NLM and the U.S.          
*  Government do not and cannot warrant the performance or results that    
*  may be obtained by using this software or data. The NLM and the U.S.    
*  Government disclaim all warranties, express or implied, including       
*  warranties of performance, merchantability or fitness for any particular
*  purpose.                                                                
*                                                                          
*  Please cite the author in any work or product based on this material.   
*
* ===========================================================================
*
* RCS $Id: asndhuff.c,v 6.1 1998/06/12 19:27:02 kans Exp $
*
* Author:  Greg Schuler
*
* Version Creation Date: 9/23/92
*
* File Description:
	asndhuff  --  decompresses a compressed ASN,1 (CASN) file.

	A Simple program to demonstrate the functions in casn.c
*
* Modifications:  
* --------------------------------------------------------------------------
* Date      Name        Description of modification
* --------  ----------  -----------------------------------------------------
* 03-22-95  Schuler     Added cases for all current doc types
* 03-22-95  Schuler     Use MonitorPtr instead of Handles for Monitors
*
* ==========================================================================
*/

#include <casn.h>

#define REVISION_STR "$Revision: 6.1 $"

void do_medline PROTO((CASN_Handle casnh, AsnIoPtr aout));
void do_sequence PROTO((CASN_Handle casnh, AsnIoPtr aout));
void stream_medline PROTO((AsnIoPtr ain));
void stream_sequence PROTO((AsnIoPtr ain));

#define NUMARGS 2
Args myargs[NUMARGS] = {
	{ "Input file", NULL, NULL, NULL, FALSE, 'i', ARG_FILE_IN, 0.0,0,NULL},
	{ "Output file", NULL, NULL, NULL, TRUE, 'o', ARG_FILE_OUT, 0.0,0,NULL}};


Int2 Main (void)
{
	CASN_Handle casnh;
	AsnIoPtr     aout, ain;

	if (! GetArgs("AsnDHuff 1.0", NUMARGS, myargs))
		return 1;

	if (! SeqEntryLoad())
	{
		ErrShow();
		return 1;
	}

	if ((casnh = CASN_Open(myargs[0].strvalue)) == NULL)
	{
		ErrShow();
		return 1;
	}

	/*****************************************************************
	*
	*   this part reads complete entries from compressed file and writes
	*   them to an uncompressed file
	*     if output file not given, it skips this step
	******************************************************************/

	if (myargs[1].strvalue != NULL)
	{
		if ((aout = AsnIoOpen(myargs[1].strvalue,"wb")) ==NULL)
		{
			Message(MSG_FATAL, "Can't create %s\n", myargs[1].strvalue);
			return 1;
		}

		switch (CASN_DocType(casnh))
		{
			case CASN_TypeMed :  /* obsolete */
			case CASN_Type_ml :
				do_medline(casnh,aout);
				break;
		
			case CASN_TypeSeq :  /* obsolete */
			case CASN_Type_aa :
			case CASN_Type_nt :
				do_sequence(casnh,aout);
				break;

			default :
				Message(MSG_ERROR,"Unknown document type");
				break;
		}

		AsnIoClose(aout);

		CASN_Close(casnh);     /* close and reopen file */
		if ((casnh = CASN_Open(myargs[0].strvalue)) == NULL)
		{
			ErrShow();
			return 1;
		}
	}

	
	/*************************************************************************
	*
	*   This part processes the compressed file as a regular asn stream
	*   displaying title lines for sequence or medline records.
	*     if output file IS given, it skips this step
	*
	*************************************************************************/

	if (myargs[1].strvalue == NULL)
	{
		ain = CASN_GetAsnIoPtr(casnh);    /* get the stream pointer */

		switch (CASN_DocType(casnh))
		{
			case CASN_TypeMed :  /* obsolete */
			case CASN_Type_ml :
				stream_medline(ain);
				break;
	
			case CASN_TypeSeq :  /* obsolete */
			case CASN_Type_aa :
			case CASN_Type_nt :
				stream_sequence(ain);
				break;

			default :
				Message(MSG_ERROR,"Unknown document type");
				break;
		}
	}


	CASN_Close(casnh);

	return 0;
}


void do_medline (CASN_Handle casnh, AsnIoPtr aout)
{
	MedlineEntryPtr entry;
	DataVal val;
	Int4 total, count;
	AsnModulePtr amp = AsnAllModPtr();
	AsnTypePtr typePubSet = AsnTypeFind(amp,"Pub-set");
	AsnTypePtr typeMedline = AsnTypeFind(amp,"Pub-set.medline");
	AsnTypePtr typeMedlineE = AsnTypeFind(amp,"Pub-set.medline.E");
	Monitor *pmonitor;

	total = CASN_DocCount(casnh);

	pmonitor = MonitorIntNew("MEDLINE Entries Decompressed", 0, total);

	AsnWrite(aout,typePubSet,&val);
	AsnStartStruct(aout,typeMedline);

	for (count=0; entry=CASN_NextMedlineEntry(casnh); ++count)
	{
		MedlineEntryAsnWrite(entry,aout,typeMedlineE);
		MedlineEntryFree(entry);
		MonitorIntValue(pmonitor, count+1);
	}

	AsnEndStruct(aout,typeMedline);
	MonitorFree(pmonitor);
	if (count != total)
		ErrPost(1,1, "Only %ld of %ld records read", count, total);
	return;
}


void do_sequence (CASN_Handle casnh, AsnIoPtr aout)
{
	SeqEntryPtr entry;
	DataVal val;
	Int4 total, count;
	AsnModulePtr amp = AsnAllModPtr();
	AsnTypePtr typeSet = AsnTypeFind(amp,"Bioseq-set");
	AsnTypePtr typeSetSet = AsnTypeFind(amp,"Bioseq-set.seq-set");
	AsnTypePtr typeSetSetE = AsnTypeFind(amp,"Bioseq-set.seq-set.E");
	Monitor *pmonitor;

	if (typeSet==NULL || typeSetSet==NULL || typeSetSetE==NULL)
	{
		Message(MSG_ERROR, "one or more atp's is NULL\n");
		return;
	}

	total = CASN_DocCount(casnh);

	pmonitor = MonitorIntNew("SEQUENCE Entries Decompressed", 0, total);

	AsnStartStruct(aout,typeSet);
	AsnStartStruct(aout,typeSetSet);

	for (count=0; entry=CASN_NextSeqEntry(casnh); ++count)
	{
		SeqEntryAsnWrite(entry,aout,typeSetSetE);
		SeqEntryFree(entry);
		MonitorIntValue(pmonitor, count+1);
	}

	AsnEndStruct(aout,typeSetSet);
	AsnEndStruct(aout,typeSet);

	MonitorFree(pmonitor);

	if (count != total)
		ErrPost(1,1, "Only %ld of %ld records read", count, total);
	return;
}

void stream_medline (AsnIoPtr ain)
{
	AsnTypePtr atp;
	DataVal val;
	AsnModulePtr amp = AsnAllModPtr();
	AsnTypePtr typePubSet = AsnTypeFind(amp,"Medline-entry");
	AsnTypePtr title = AsnTypeFind(amp,"Title.E.name");
	Monitor *pmonitor;

	pmonitor = MonitorStrNew("MEDLINE Titles Decompressed", 60);

	atp = typePubSet;
	while ((atp = AsnReadId(ain, amp, atp)) != NULL)
	{
		if (atp == title)
		{
			AsnReadVal(ain, atp, &val);
			MonitorStrValue(pmonitor, (CharPtr)val.ptrvalue);
			MemFree(val.ptrvalue);
		}
		else
			AsnReadVal(ain, atp, NULL);
	}
	MonitorFree(pmonitor);
	return;
}

void stream_sequence (AsnIoPtr ain)
{
	AsnTypePtr atp;
	DataVal val;
	AsnModulePtr amp = AsnAllModPtr();
	AsnTypePtr typeSeqEntry = AsnTypeFind(amp,"Seq-entry");
	AsnTypePtr title = AsnTypeFind(amp,"Seq-descr.E.title");
	Monitor *pmonitor;

	pmonitor = MonitorStrNew("Sequence Titles Decompressed", 60);

	atp = typeSeqEntry;
	while ((atp = AsnReadId(ain, amp, atp)) != NULL)
	{
		if (atp == title)
		{
			AsnReadVal(ain, atp, &val);
			MonitorStrValue(pmonitor, (CharPtr)val.ptrvalue);
			MemFree(val.ptrvalue);
		}
		else
			AsnReadVal(ain, atp, NULL);
		if (! AsnGetLevel(ain))       /* finished reading a Seq-entry */
			atp = typeSeqEntry;		  /* reset to start next one */
	}
	MonitorFree(pmonitor);
	return;
}



